File: fluttermultiplatform/lib/models/batch.dart
import 'shrimp_type.dart';

class Batch {
  final String id;
  final String harvestId;
  final String batchCode;
  final double quantity;
  final String status;
  final ShrimpType? shrimpType;
  final String createdAt;

  Batch({required this.id, this.harvestId = '', this.batchCode = '', this.quantity = 0, this.status = '', this.shrimpType, this.createdAt = ''});

  factory Batch.fromJson(Map<String, dynamic> json) => Batch(
        id: json['id']?.toString() ?? '',
        harvestId: json['harvest_id']?.toString() ?? '',
        batchCode: json['batch_code'] ?? '',
        quantity: (json['quantity'] ?? 0).toDouble(),
        status: json['status'] ?? '',
        shrimpType: json['shrimp_type'] != null ? ShrimpType.fromJson(json['shrimp_type']) : null,
        createdAt: json['created_at'] ?? '',
      );
}

File: backend-spring/src/main/java/com/cronos/entity/Batch.java
package com.cronos.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@Table(name = "batches")
@SQLDelete(sql = "UPDATE batches SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Batch {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "uuid")
    private UUID id;

    @JsonProperty("harvest_id")
    @Column(name = "harvest_id", nullable = false, columnDefinition = "uuid")
    private UUID harvestId;

    @JsonProperty("batch_code")
    @Column(name = "batch_code", nullable = false, unique = true, length = 50)
    private String batchCode;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal quantity;

    @Column(length = 50)
    @Builder.Default
    private String status = "available";

    @JsonProperty("created_at")
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @JsonProperty("updated_at")
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @JsonIgnore
    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "harvest_id", insertable = false, updatable = false)
    @JsonIgnore
    private Harvest harvest;

    @OneToMany(mappedBy = "batch", fetch = FetchType.LAZY)
    @JsonIgnore
    private List<Product> products;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

